shaders/src/main/glsl/samples/310es/quicksort_palette.frag

```diff
@@ -0,0 +1,121 @@
+ */
+
+precision highp float;
+
+layout(location = 0) out vec4 _GLF_color;
+
+uniform vec2 injectionSwitch;
+
+uniform vec2 resolution;
+
+struct QuicksortObject{
+    int numbers[10];
+};
+
+QuicksortObject obj;
+
+ void swap(int i, int j) {
+     int temp = obj.numbers[i];
+     obj.numbers[i] = obj.numbers[j];
+     obj.numbers[j] = temp;
+ }
+
+ // Since "partition" is the preserved word, we add prefix to this function name to prevent an error.
+ int performPartition(int l, int h) {
+     // The rightmost element is chosen as a pivot.
+     int pivot = obj.numbers[h];
+     int i = (l - 1);
+  
+     for (int j = l; j <= h - 1; j++) {
+         if (obj.numbers[j] <= pivot) {
+             i++;
+             swap(i, j);
+         }
+     }
+     swap(i + 1, h);
+     return (i + 1);
+ }
+
+ void quicksort() {
+     int l = 0, h = 9;
+     int stack[10];
+     int top = -1;
+  
+     stack[++top] = l;
+     stack[++top] = h;
+  
+     while (top >= 0) {
+         h = stack[top--];
+         l = stack[top--];
+  
+         int p = performPartition(l, h);
+  
+         if (p - 1 > l) {
+             stack[++top] = l;
+             stack[++top] = p - 1;
+         }
+         if (p + 1 < h) {
+             stack[++top] = p + 1;
+             stack[++top] = h;
+         }
+     }
+ }
+
+vec3 palette(float t, vec3 a, vec3 b, vec3 c, vec3 d) {
+    return mix(c, d + d, a - b);
+}
+
+void main() {
+    // Initialize decreasing values to an array item starting from 10.
+    for (int i = int(injectionSwitch.x); i < 10; i ++) {
+      obj.numbers[i] = (10 - i) * int(injectionSwitch.y);
+    }
+    quicksort();
+    vec2 uv = gl_FragCoord.xy / resolution;
+
+    vec3 color = palette(uv.x, vec3( float(obj.numbers[7]) * 0.1 ), vec3(0.5, float(obj.numbers[0]) * 0.1, 0.8), trunc(vec3(injectionSwitch.y)), vec3(injectionSwitch.x, 0.3, 0.7));
+    color.x += step(sin((uv.y -  0.1 + abs(uv.x) * 0.5) * 250.0), 0.0);
+    if (uv.x > (1.0/4.0)) {
+      int count = int(injectionSwitch.x);
+      do {
+        color += palette(uv.x, vec3(0.5, float(obj.numbers[8]) * 0.1, 0.2), vec3(0.5), trunc(vec3(injectionSwitch.y)), vec3(float(obj.numbers[4]) * 0.1, injectionSwitch.x, 0.6));
+        color += sin(gl_FragCoord.x * cos( 1.0 / 15.0 ) * 10.0 ) + cos( gl_FragCoord.y * cos( 1.0 / 15.0 ) * 10.0 );
+        count++;
+      } while (count != obj.numbers[int(injectionSwitch.x)]);
+    }
+    if (uv.x > (2.0/4.0)) {
+      int count = int(injectionSwitch.x);
+      do {
+        color -= palette(trunc(uv.x), vec3(float(obj.numbers[4]) * 0.1), trunc(vec3(0.1)), vec3(float(obj.numbers[int(injectionSwitch.y)]) * 0.1), vec3(injectionSwitch.x, float(obj.numbers[2]) * 0.1 , float(obj.numbers[8]) * 0.1));
+        color.z += sin( gl_FragCoord.y * sin( 1.0 / 10.0 ) * 40.0 ) + cos( gl_FragCoord.x * sin( 1.0 / 25.0 ) * 40.0 );
+        count++;
+      } while (count != obj.numbers[1]);
+    }
+    if(uv.x > (3.0/4.0)) {
+      int count = int(injectionSwitch.x);
+      do {
```

Comment:
I am concerned that this might give nondetermnistic results in the case that one of the components of the result of palette is close to or equal to 0.0, due to the division.  I am seeing disagreement between Intel Mesa results and results for other platforms, and the difference seems to be related to this line of code.

Can you either avoid division here, or do something to ensure that the divisors will not be close to zero (if indeed they can be currently)?